hash_practice/exercises.py

```diff
@@ -1,19 +1,54 @@
 
+from tabnanny import check
+
+
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n*m log m) 
+        Space Complexity: O(n) since a hash table was created
     """
-    pass
+    group_anagrams = dict()
+    for word in strings: #O(n) where n is the length of input
+        sorted_ver = "".join(sorted(word)) #O(m log m) where m is the length of each word
+        if sorted_ver not in group_anagrams:
+            group_anagrams[sorted_ver] = [word]
+        else:
+            group_anagrams[sorted_ver].append(word)
+
+    output = []
+    for word_list in group_anagrams.values(): #O(n)
+        output.append(word_list)
+    return output
+        
 
 def top_k_frequent_elements(nums, k):
     """ This method will return the k most common elements
         In the case of a tie it will select the first occuring element.
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(m*n)
+        Space Complexity: O(n) for num_freq
     """
-    pass
+    if not nums:
+        return nums
+
+    num_freq = dict() #O(n) for space
+    
+    for num in nums: #O(n) for time
+        if num not in num_freq:
+            num_freq[num] = 1
+        else:
+            num_freq[num] += 1
+
+    output = []
+    for i in range(k): #O(m) for time
+        if len(output) == k:
+            return output
+        max_value = max(num_freq.values())
+        for key,val in list(num_freq.items()): #O(n) for time where n is the length of nums, also O(n) to cast into list
+            if val == max_value:
+                output.append(key)
+                del num_freq[key]
+    return output
 
 
 def valid_sudoku(table):

```

Comment:
✨ Since the table is of constant size (always 9 x 9), we can say this has O(1) time and space complexity!